hera_sim/tests/test_compare_pyuvsim.py

```diff
@@ -0,0 +1,331 @@
+    # ---------------------------------------------------------------------------
+    uvd_uvsim = uvsim.run_uvdata_uvsim(
+        uvdata,
+        BeamList(beams),
+        beam_dict=beam_dict,
+        catalog=simsetup.SkyModelData(sky_model),
+        quiet=True,
+    )
+
+    # ---------------------------------------------------------------------------
+    # Compare results
+    # ---------------------------------------------------------------------------
+    # Set relative/absolute tolerances depending on no. of sources
+    # (N.B. vis_cpu source position correction approximation degrades with time)
+    if nsource < 10:
+        # Very stringent for a few sources
+        rtol = 1e-4
+        atol = 1e-7
+    else:
+        # Within 0.1% or so for many sources
+        rtol = 1e-3
+        atol = 1e-5
+
+    # Loop over baselines and compare
+    diff_re = 0.0
+    diff_im = 0.0
+    for i in range(nants):
+        for j in range(i, nants):
+
+            # Get visibilities for this baseline and polarisation
+            d_uvsim = uvd_uvsim.get_data((i, j, "xx")).T  # pyuvsim visibility
+            d_viscpu = uvd_viscpu.get_data((i, j, "xx")).T  # vis_cpu visibility
+
+            # Print diagnostics
+            print("\nBaseline:", i, j)
+            print("V_uvsim: ", d_uvsim)
+            print("V_viscpu:", d_viscpu)
+            print("Ratio of abs():", np.abs(d_uvsim) / np.abs(d_viscpu))
+            print("Ratio of real: ", np.real(d_uvsim) / np.real(d_viscpu))
+            print("Ratio of imag: ", np.imag(d_uvsim) / np.imag(d_viscpu))
+
+            # Keep track of maximum difference
+            delta = d_uvsim - d_viscpu
+            if np.abs(np.max(delta.real)) > diff_re:
+                diff_re = np.abs(np.max(delta.real))
+            if np.abs(np.max(delta.imag)) > diff_im:
+                diff_im = np.abs(np.max(delta.imag))
+
+            # Very stringent threshold of 0.1%
+            assert np.allclose(
+                d_uvsim.real,
+                d_viscpu.real,
+                rtol=rtol,
+                atol=atol,
+            ), "Max. difference (re): %10.10e" % (diff_re,)
+            assert np.allclose(
+                d_uvsim.imag,
+                d_viscpu.imag,
+                rtol=rtol,
+                atol=atol,
+            ), "Max. difference (im): %10.10e" % (diff_im,)
+
+
+def test_compare_1src_gauss():
+    # Single source, Gaussian beam, unpolarized
+    compare_viscpu_with_pyuvsim(nsource=1, beam_type="gaussian")
+
+
+def test_compare_1src_polybeam():
+    # Single source, PolyBeam beam, unpolarized
+    compare_viscpu_with_pyuvsim(nsource=1, beam_type="PolyBeam")
+
+
+def test_compare_1src_polybeam_pol():
+    # Single source, PolyBeam beam, polarized
+    compare_viscpu_with_pyuvsim(nsource=1, beam_type="PolyBeam polarized")
+
+
+def test_compare_100src_gauss():
+    # Single source, Gaussian beam, unpolarized
+    compare_viscpu_with_pyuvsim(nsource=100, beam_type="gaussian")
```

Comment:
It is more idiomatic to write these as parameterizations, so instead of:

```python
def compare_viscpu_with_pyuvsim(nsource, beam_type):
```

you'd have

```python
@pytest.mark.parameterize('nsource,beam_type', [(1, 'gaussian'), (1, 'PolyBeam'), ...])
def test_viscpu_vs_pyuvsim(nsource, beam_type):
```